modules/rostests/kmtests/kmtest_drv/kmtest_fsminifilter.c

```diff
@@ -0,0 +1,521 @@
+                                   OBJ_KERNEL_HANDLE | OBJ_CASE_INSENSITIVE,
+                                   NULL,
+                                   SecurityDescriptor);
+
+
+        /* Create the usermode communication port */
+        Status = FltCreateCommunicationPort(FilterData.Filter,
+                                            &FilterData.ServerPort,
+                                            &ObjectAttributes,
+                                            NULL,
+                                            FilterConnect,
+                                            FilterDisconnect,
+                                            FilterMessage,
+                                            MaxConnections);
+
+        /* Free the security descriptor */
+        FltFreeSecurityDescriptor(SecurityDescriptor);
+
+        if (!NT_SUCCESS(Status))
+        {
+            goto cleanup;
+        }
+    }
+
+    if (!(Flags & TESTENTRY_NO_START_FILTERING))
+    {
+        /* Start filtering the requests */
+        Status = FltStartFiltering(FilterData.Filter);
+    }
+
+cleanup:
+    if (!NT_SUCCESS(Status))
+    {
+        if (FilterData.ServerPort)
+        {
+            FltCloseCommunicationPort(FilterData.ServerPort);
+        }
+        if (FilterData.Filter)
+        {
+            FltUnregisterFilter(FilterData.Filter);
+        }
+    }
+
+    return Status;
+}
+
+/**
+ * @name DriverUnload
+ *
+ * Driver cleanup funtion.
+ *
+ * @param Flags
+ *        Flags describing the unload request
+ */
+NTSTATUS
+FLTAPI
+FilterUnload(
+    _In_ FLT_FILTER_UNLOAD_FLAGS Flags)
+{
+    PAGED_CODE();
+    UNREFERENCED_PARAMETER(Flags);
+
+    DPRINT("DriverUnload\n");
+
+    TestFilterUnload(Flags);
+
+    /* Close the port and unregister the filter */
+    if (FilterData.ServerPort)
+    {
+        FltCloseCommunicationPort(FilterData.ServerPort);
+    }
+    if (FilterData.Filter)
+    {
+        FltUnregisterFilter(FilterData.Filter);
+    }
+
+    if (Callbacks)
+    {
+        ExFreePoolWithTag(Callbacks, KMTEST_FILTER_POOL_TAG);
+        Callbacks = NULL;
+    }
+
+    return STATUS_SUCCESS;
+}
+
+
+/**
+ * @name FilterInstanceSetup
+ *
+ * Volume attach routine
+ *
+ * @param FltObjects
+ *        Filter Object Pointers
+ *        Pointer to an FLT_RELATED_OBJECTS structure that contains opaque pointers
+ *        for the objects related to the current operation
+ * @param Flags
+ *        Bitmask of flags that indicate why the instance is being attached
+ * @param VolumeDeviceType
+ *        Device type of the file system volume
+ * @param VolumeFilesystemType
+ *        File system type of the volume
+ *
+ * @return Status.
+ *         Return STATUS_SUCCESS to attach or STATUS_FLT_DO_NOT_ATTACH to ignore
+ */
+NTSTATUS
+FLTAPI
+FilterInstanceSetup(
+    _In_ PCFLT_RELATED_OBJECTS FltObjects,
+    _In_ FLT_INSTANCE_SETUP_FLAGS Flags,
+    _In_ DEVICE_TYPE VolumeDeviceType,
+    _In_ FLT_FILESYSTEM_TYPE VolumeFilesystemType)
+{
+#if 0
```

Comment:
Might want to explain why this is commented out

Reply:
Yeah, I've not implemented FltGetVolumeProperties yet :/